R/execute.R

```diff
@@ -95,29 +98,47 @@ execute_array_udf <- function(namespace, array, udf, selectedRanges, layout=NULL
   queryRanges <- QueryRanges$new(layout=layout, ranges=selectedRanges)
   multi_array_udf$ranges = queryRanges
 
+  if (!is.null(result_format)) {
+    multi_array_udf$result_format <- ResultFormat$new(result_format)
+  }
+
   # Attrs can be optionally specified by the client. If they are not, the
   # server-side code will load all attributes.
   if (!is.null(attrs)) {
     multi_array_udf$buffers <- attrs
   }
 
   # Make the network request.
-  resultObject <- udfapi$SubmitUDF(namespace=namespace, array=array, udf=multi_array_udf)
+  result <- udfapi$SubmitUDF(namespace=namespace, array=array, udf=multi_array_udf)
 
   # Decode the results.
-  if (typeof(resultObject) != "raw") {
-    className <- class(resultObject)[1]
+  if (typeof(result) != "raw") {
+    className <- class(result)[1]
     if (className == "ApiResponse") {
-      stop("tiledbcloud: received error response: ", resultObject$content, call.=FALSE)
+      stop("tiledbcloud: received error response: ", result$content, call.=FALSE)
     } else {
-      stop("tiledbcloud: received error response: ", class(resultObject)[1], call.=FALSE)
+      stop("tiledbcloud: received error response: ", class(result)[1], call.=FALSE)
     }
   }
   # This is a serialized R object sent from the server-side R code, passed
   # through as-is by the REST server (opaquely to it), to the client-side R
   # code (us).
-  resultString <- rawToChar(resultObject)
-  resultJSON <- jsonlite::fromJSON(resultString)
-  resultValue <- resultJSON$value
-  resultValue
+
+  formattedResult <- NULL
+  switch(result_format,
+    native={
+      formattedResult <- unserialize(result, NULL)
+    },
+    json={
+      resultString <- rawToChar(result)
+      resultJSON <- jsonlite::fromJSON(resultString)
+      formattedResult <- resultJSON$value
+    },
+    arrow={
+      formattedResult <- arrow::read_ipc_stream(result)
```

Comment:
Decision point: @aaronwolen @eddelbuettel do we want to decode here, or return the result as-is to the calling script so they can `arrow::read_ipc_stream` as they wish?

Reply:
Who is "they" here?

Reply:
Sorry -- people using `TileDB-Cloud-R`

Reply:
For the record: off-line conversation with @Shelnutt2 : this parameter indeed controls the wire-protocol used between the REST server and this library; in all cases the return value back to the user should be deserialized -- they do not receive JSON or Arrow. This is intentional and consistent with the `TileDB-Cloud-Py` library. If we decide to later add a flag allowing return-to-caller of JSON or Arrow, that will be a separate feature, implemented for Python as well as R.